File: server-python/controllers/candidates_controller.py
from database.db import Session, Candidate, Test, TestCandidate

def get_all_candidates():
    """
    Get all candidates with their assigned tests.
    
    Returns:
        list: List of candidate dictionaries
    """
    session = Session()
    try:
        candidates = session.query(Candidate).all()
        return [candidate.to_dict() for candidate in candidates]
    except Exception as e:
        raise Exception(f"Error fetching candidates: {str(e)}")
    finally:
        session.close()

def get_candidate_by_id(candidate_id):
    """
    Get a specific candidate by ID.
    
    Args:
        candidate_id (int): The candidate ID
        
    Returns:
        dict: Candidate details
    """
    session = Session()
    try:
        candidate = session.query(Candidate).filter(Candidate.id == candidate_id).first()
        if not candidate:
            raise Exception("Candidate not found")
        
        return candidate.to_dict()
    except Exception as e:
        raise Exception(f"Error fetching candidate: {str(e)}")
    finally:
        session.close()

def update_candidate_status(candidate_id, completed):
    """
    Update a candidate's completion status.
    
    Args:
        candidate_id (int): The candidate ID
        completed (bool): New completion status
        
    Returns:
        dict: Updated candidate details
    """
    session = Session()
    try:
        candidate = session.query(Candidate).filter(Candidate.id == candidate_id).first()
        if not candidate:
            raise Exception("Candidate not found")
        
        candidate.completed = completed
        session.commit()
        
        return candidate.to_dict()
    except Exception as e:
        session.rollback()
        raise Exception(f"Error updating candidate: {str(e)}")
    finally:
        session.close() 

File: server/controllers/candidatesController.js
const { db } = require('../database/db');

// Get all candidates
exports.getAllCandidates = async (req, res) => {
  try {
    // Get all candidates with their assigned tests
    const candidates = await db('candidates')
      .select('candidates.*')
      .orderBy('candidates.id');
    
    // For each candidate, get their tests
    for (let candidate of candidates) {
      const testAssignments = await db('test_candidates')
        .join('tests', 'test_candidates.test_id', '=', 'tests.id')
        .where('test_candidates.candidate_id', candidate.id)
        .select('tests.name', 'tests.id', 'test_candidates.completed');
      
      candidate.testsAssigned = testAssignments.map(test => ({
        id: test.id,
        name: test.name,
        completed: test.completed
      }));
    }
    
    res.json(candidates);
  } catch (error) {
    console.error('Error fetching candidates:', error);
    res.status(500).json({ error: 'Failed to fetch candidates' });
  }
};

// Get a specific candidate by ID
exports.getCandidateById = async (req, res) => {
  const { id } = req.params;
  
  try {
    const candidate = await db('candidates')
      .where('id', id)
      .first();
    
    if (!candidate) {
      return res.status(404).json({ error: 'Candidate not found' });
    }
    
    // Get tests assigned to this candidate
    const testAssignments = await db('test_candidates')
      .join('tests', 'test_candidates.test_id', '=', 'tests.id')
      .where('test_candidates.candidate_id', id)
      .select('tests.name', 'tests.id', 'test_candidates.completed');
    
    candidate.testsAssigned = testAssignments.map(test => ({
      id: test.id,
      name: test.name,
      completed: test.completed
    }));
    
    res.json(candidate);
  } catch (error) {
    console.error('Error fetching candidate:', error);
    res.status(500).json({ error: 'Failed to fetch candidate' });
  }
};

// Update candidate completion status
exports.updateCandidateStatus = async (req, res) => {
  const { id } = req.params;
  const { completed } = req.body;
  
  try {
    const updatedCandidate = await db('candidates')
      .where('id', id)
      .update({ completed }, ['id', 'name', 'email', 'completed']);
    
    if (!updatedCandidate) {
      return res.status(404).json({ error: 'Candidate not found' });
    }
    
    res.json(updatedCandidate[0]);
  } catch (error) {
    console.error('Error updating candidate:', error);
    res.status(500).json({ error: 'Failed to update candidate' });
  }
}; 